src/CMakeLists.txt

```diff
@@ -1295,63 +1295,92 @@ add_subdirectory(util)
 # LIBCVC5_GEN_SRCS (via libcvc5_add_sources). We can now build libcvc5.
 
 set_source_files_properties(${LIBCVC5_GEN_SRCS} PROPERTIES GENERATED TRUE)
-add_library(cvc5 ${LIBCVC5_SRCS} ${LIBCVC5_GEN_SRCS}
-  $<TARGET_OBJECTS:cvc5base> $<TARGET_OBJECTS:cvc5context>)
-target_include_directories(cvc5
+
+# First build the object library
+add_library(cvc5-obj OBJECT ${LIBCVC5_SRCS} ${LIBCVC5_GEN_SRCS})
+target_compile_definitions(cvc5-obj PRIVATE -D__BUILDING_CVC5LIB)
+set_target_properties(cvc5-obj PROPERTIES POSITION_INDEPENDENT_CODE ON)
+target_link_libraries(cvc5-obj PUBLIC $<TARGET_OBJECTS:cvc5base> $<TARGET_OBJECTS:cvc5context>)
+# add_dependencies() is necessary for cmake versions before 3.21
+add_dependencies(cvc5-obj cvc5base cvc5context)
+# Add libcvc5 dependencies for generated sources.
+add_dependencies(cvc5-obj gen-expr gen-gitinfo gen-options gen-tags gen-theory)
+
+# Link the shared library
+add_library(cvc5-shared SHARED)
+target_link_libraries(cvc5-shared PRIVATE cvc5-obj)
+set_target_properties(cvc5-shared PROPERTIES SOVERSION ${CVC5_SOVERSION})
+set_target_properties(cvc5-shared PROPERTIES OUTPUT_NAME cvc5)
+set_target_properties(cvc5-shared PROPERTIES POSITION_INDEPENDENT_CODE ON)
+target_include_directories(cvc5-shared
   PUBLIC
     $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
     $<INSTALL_INTERFACE:include>
 )
-
-include(GenerateExportHeader)
-generate_export_header(cvc5)
-
-install(TARGETS cvc5
+install(TARGETS cvc5-shared
   EXPORT cvc5-targets
   LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
-  ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
+  ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+)
 
-set_target_properties(cvc5 PROPERTIES SOVERSION ${CVC5_SOVERSION})
-target_compile_definitions(cvc5 PRIVATE -D__BUILDING_CVC5LIB)
-# Add libcvc5 dependencies for generated sources.
-add_dependencies(cvc5 gen-expr gen-gitinfo gen-options gen-tags gen-theory)
+if(ENABLE_STATIC_LIBRARY)
+  add_library(cvc5-static STATIC)
+  target_link_libraries(cvc5-static PRIVATE cvc5-obj)
+  set_target_properties(cvc5-static PROPERTIES OUTPUT_NAME cvc5)
+  target_include_directories(cvc5-static
+    PUBLIC
+      $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
+      $<INSTALL_INTERFACE:include>
+  )
+  install(TARGETS cvc5-obj cvc5-static
```

Comment:
Why is `cvc5-obj` an install target?

Reply:
This is an excellent question, that I have no really good answer to.
If we remove it, cmake will complain:
```
CMake Error: install(EXPORT "cvc5-targets" ...) includes target "cvc5-static" which requires target "cvc5-obj" that is not in any export set.
```
even though, `cvc5-static` only links against `cvc5-obj` with `PRIVATE`. This is one of the quirks I had in mind when I said something to the effect of "we support static builds, but they may behave in weird ways when someone tries to install them".

What I *think* is going on here: when linking statically, you need to explicitly link against all dependencies, and their dependencies, etc. To properly reflect this in the export, all these targets must be in the export set. What I don't understand, though, is why cmake does not enforce the same for the recursive dependencies. `cvc5Targets.cmake` also contains references of `CaDiCaL` or `SymFPU`, and thus actually using `cvc5-static` would most likely fail anyway...

Reply:
This link might help to resolve the issue: https://discourse.cmake.org/t/install-ing-a-static-library-also-requires-installing-an-internal-object-library/683/3

Reply:
Passing cvc5-obj is expected behavior.